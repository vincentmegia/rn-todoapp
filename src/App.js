import React, { useCallback, useEffect, useState } from "react"
import {
   createTheme,
   ThemeProvider,
   Avatar,
   Icon,
   Text,
   Image,
} from "@rneui/themed"
import { SafeAreaView, ScrollView, View } from "react-native"
import styles from "./style"
import Task from "./components/task"
import { launchImageLibrary } from "react-native-image-picker"
import Add from "./components/add"
import {
   getImage,
   getTasks,
   saveImage,
   saveTasks,
} from "./repository/storageRepository"
import images from "./assets"

const theme = createTheme({
   components: {
      Button: {
         raised: true,
      },
   },
})

/**
 *
 * @returns
 */
const App = () => {
   const [imagePath, setImagePath] = useState({})
   const [tasks, setTasks] = useState(new Map())
   const [overlay, setOverlay] = useState(false)
   const [error, setError] = useState("")

   useEffect(() => {
      const fetchProfilePic = async () => {
         const profilePic = await getImage()
         if (profilePic) {
            setImagePath({
               isImageAvailable: true,
               profilePic: JSON.parse(profilePic),
            })
         }
      }

      const fetchTasks = async () => {
         const result = await getTasks()
         setTasks(new Map(JSON.parse(result)))
      }

      fetchProfilePic().catch((profileError) => {
         setError("an error has occured: ", profileError)
      })
      fetchTasks().catch((tasksError) => {
         setError("an error has occured: ", tasksError)
      })
   }, [])

   const onAvatarPress = useCallback(async () => {
      var options = {
         storageOptions: {
            skipBackup: true,
            path: "images",
         },
      }

      await launchImageLibrary(options, (response) => {
         if (response.didCancel || response.error || response.customButton) {
            console.log("User cancelled image picker")
            return
         }

         const source = response.assets[0].uri
         saveImage(source)
         setImagePath({
            isImageAvailable: true,
            profilePic: source,
         })
      })
   }, [])

   const onTaskCompleted = useCallback(
      (item) => {
         tasks.set(item.key, item.task)
         saveTasks(new Map(tasks))
      },
      [tasks]
   )

   const onNewPress = useCallback(() => {
      setOverlay(true)
   }, [])

   const onOverlayButtonPress = useCallback(() => {
      setOverlay(false)
   }, [])

   const onDelete = useCallback(
      (item) => {
         tasks.delete(item.key)
         setTasks(new Map(tasks))
         saveTasks(new Map(tasks))
      },
      [tasks]
   )

   const onModalButtonPress = useCallback(
      (text) => {
         setOverlay(false)
         if (!text) {
            return
         }

         tasks.set(new Date().getMilliseconds(), {
            text,
            status: false,
         })
         setTasks(new Map(tasks))
         saveTasks(new Map(tasks))
      },
      [tasks]
   )

   return (
      <SafeAreaView style={{ flex: 1 }}>
         <ThemeProvider theme={theme}>
            <Avatar
               size="xlarge"
               rounded
               title="Hi"
               source={
                  imagePath.isImageAvailable
                     ? { uri: imagePath.profilePic }
                     : { uri: undefined }
               }
               containerStyle={styles.avatar}
            >
               <Avatar.Accessory onPress={onAvatarPress} size={30} />
            </Avatar>
            <View style={styles.add}>
               <Icon name="add" raised onPress={onNewPress} />
            </View>
            <Add
               visible={overlay}
               onButtonPress={onModalButtonPress}
               onBackdropPress={onOverlayButtonPress}
            />
            {tasks.size === 0 && (
               <View style={styles.emptyContainer}>
                  <Text style={styles.boaredText}>
                     Let's do it!. I was really bored when I wrote this app.
                     Good test for M2 chip
                  </Text>
                  <View style={styles.penContainer}>
                     <Image source={images.pen} style={styles.penImage} />
                  </View>
                  <Text style={[styles.penText]}>
                     Add new tasks via clicking the plus sign (+)
                  </Text>
               </View>
            )}
            {tasks.size > 0 && (
               <View style={styles.taskContainer}>
                  <ScrollView contentContainerStyle={styles.scrollView}>
                     {tasks &&
                        Array.from(tasks).map((task, index) => (
                           <Task
                              key={index}
                              item={{ key: task[0], task: task[1] }}
                              onTaskCompleted={onTaskCompleted}
                              onDelete={onDelete}
                           />
                        ))}
                  </ScrollView>
               </View>
            )}
         </ThemeProvider>
      </SafeAreaView>
   )
}

export default App
