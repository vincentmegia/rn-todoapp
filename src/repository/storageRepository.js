import AsyncStorage from '@react-native-community/async-storage'

export const saveTasks = async map => {
  try {
    await AsyncStorage.setItem('tasks', JSON.stringify(Array.from(map)))
  } catch (error) {
    console.log('error occurred from saveTasks:', JSON.stringify(error))
  }
}

export const getTasks = async () => {
  try {
    return await AsyncStorage.getItem('tasks')
  } catch (err) {
    console.log('error occurred from getTasks: ', JSON.stringify(err))
  }
}

export const getImage = async () => {
  try {
    return await AsyncStorage.getItem('profilePic')
  } catch (err) {
    console.log('error occurred from getImage: ', JSON.stringify(err))
  }
}

export const saveImage = async source => {
  try {
    return await AsyncStorage.setItem('profilePic', JSON.stringify(source))
  } catch (err) {
    console.log('error occurred from getImage: ', JSON.stringify(err))
  }
}
