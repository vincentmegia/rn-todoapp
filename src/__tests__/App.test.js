import React from 'react'
import App from '../App'
import {render} from '@testing-library/react-native'
import * as repository from '../repository/storageRepository'
import {act} from 'react-test-renderer'

describe('App container tests', () => {
  it('it matches snapshot', () => {
    const map = new Map()
    map.set(new Date().getMilliseconds(), {text: 'asdf', status: true})
    jest
      .spyOn(repository, 'getTasks')
      .mockReturnValue(JSON.stringify(Array.from(map)))

    jest.spyOn(repository, 'getImage').mockReturnValue('file:///fakefile.jpg')

    const {toJSON} = render(<App />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('it getImage and getTasks throws error', () => {
    jest.spyOn(repository, 'getImage').mockImplementation(() => {
      throw 'simulate error'
    })

    jest.spyOn(repository, 'getTasks').mockImplementation(() => {
      throw 'simulate error'
    })

    render(<App />)
  })
})
