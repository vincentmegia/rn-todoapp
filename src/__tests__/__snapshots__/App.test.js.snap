// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App container tests it matches snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    replaceTheme={[Function]}
    style={
      Object {
        "backgroundColor": "grey",
        "borderRadius": 75,
        "height": 150,
        "marginHorizontal": 24,
        "marginVertical": 24,
        "width": 150,
      }
    }
    theme={
      Object {
        "colors": Object {
          "background": "#ffffff",
          "black": "#242424",
          "disabled": "hsl(208, 8%, 90%)",
          "divider": "#bcbbc1",
          "error": "#ff190c",
          "grey0": "#393e42",
          "grey1": "#43484d",
          "grey2": "#5e6977",
          "grey3": "#86939e",
          "grey4": "#bdc6cf",
          "grey5": "#e1e8ee",
          "greyOutline": "#bbb",
          "platform": Object {
            "android": Object {
              "error": "#f44336",
              "grey": "rgba(0, 0, 0, 0.54)",
              "primary": "#2196f3",
              "searchBg": "#dcdce1",
              "secondary": "#9C27B0",
              "success": "#4caf50",
              "warning": "#ffeb3b",
            },
            "default": Object {
              "error": "#ff3b30",
              "grey": "#7d7d7d",
              "primary": "#007aff",
              "searchBg": "#dcdce1",
              "secondary": "#5856d6",
              "success": "#4cd964",
              "warning": "#ffcc00",
            },
            "ios": Object {
              "error": "#ff3b30",
              "grey": "#7d7d7d",
              "primary": "#007aff",
              "searchBg": "#dcdce1",
              "secondary": "#5856d6",
              "success": "#4cd964",
              "warning": "#ffcc00",
            },
            "web": Object {
              "error": "#ff190c",
              "grey": "#393e42",
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#ca71eb",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          },
          "primary": "#2089dc",
          "searchBg": "#303337",
          "secondary": "#03dac4",
          "success": "#52c41a",
          "warning": "#faad14",
          "white": "#ffffff",
        },
        "mode": "light",
        "spacing": Object {
          "lg": 12,
          "md": 8,
          "sm": 4,
          "xl": 24,
          "xs": 2,
        },
      }
    }
    updateTheme={[Function]}
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        Object {
          "backgroundColor": "transparent",
          "borderRadius": 75,
          "flex": 1,
          "overflow": "hidden",
          "position": "relative",
        }
      }
    >
      <Image
        borderRadius={75}
        onLoad={[Function]}
        source={
          Object {
            "uri": undefined,
          }
        }
        style={
          Object {
            "bottom": 0,
            "flex": 1,
            "height": undefined,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": undefined,
          }
        }
        testID="RNE__Avatar__Image"
        transitionDuration={360}
      />
      <View
        accessibilityElementsHidden={true}
        collapsable={false}
        importantForAccessibility="no-hide-descendants"
        pointerEvents="none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "flex": 1,
              "height": undefined,
              "justifyContent": "center",
              "width": undefined,
            }
          }
          testID="RNE__Image__placeholder"
        >
          <Text
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#ffffff",
                "fontSize": 75,
                "textAlign": "center",
              }
            }
          >
            Hi
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "height": undefined,
            "width": undefined,
          }
        }
        testID="RNE__Image__children__container"
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#aaa",
            "bottom": 0,
            "justifyContent": "center",
            "position": "absolute",
            "right": 0,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 1,
              "width": 1,
            },
            "shadowOpacity": 0.5,
            "shadowRadius": 2,
          },
          Object {
            "borderRadius": 15,
            "height": 30,
            "width": 30,
          },
          undefined,
        ]
      }
    >
      <View>
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
          testID="RNE__ICON__CONTAINER"
        >
          <View
            accessibilityRole="button"
            android_ripple={
              Object {
                "borderless": false,
                "color": "rgba(0, 0, 0, 0.3)",
                "radius": -5,
              }
            }
            disabled={false}
            replaceTheme={[Function]}
            testID="RNE__ICON__CONTAINER_ACTION"
            updateTheme={[Function]}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
              testID="RNE__ICON"
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#fff",
                      "fontSize": 24,
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                    Object {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
                testID="RNE__ICON__Component"
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row-reverse",
      }
    }
  >
    <View
      style={
        Object {
          "borderRadius": 28,
          "height": 52,
          "margin": 7,
          "shadowColor": "rgba(0,0,0, .4)",
          "shadowOffset": Object {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
          "width": 52,
        }
      }
      testID="RNE__ICON__CONTAINER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        replaceTheme={[Function]}
        testID="RNE__ICON__CONTAINER_ACTION"
        updateTheme={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "borderRadius": 28,
              "height": 52,
              "justifyContent": "center",
              "width": 52,
            }
          }
          testID="RNE__ICON"
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#242424",
                  "fontSize": 24,
                },
                Object {
                  "backgroundColor": "transparent",
                },
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            testID="RNE__ICON__Component"
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    replaceTheme={[Function]}
    testID="modal"
    transparent={true}
    updateTheme={[Function]}
    visible={false}
  />
  <View
    style={
      Object {
        "flex": 1,
        "marginHorizontal": 24,
        "marginTop": 80,
      }
    }
  >
    <Text
      accessibilityRole="text"
      style={
        Object {
          "color": "#242424",
          "marginLeft": "auto",
          "marginRight": "auto",
          "marginVertical": 16,
        }
      }
    >
      Let's do it!. I was really bored when I wrote this app. Good test for M2 chip
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "marginTop": 16,
        }
      }
    >
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          Object {
            "backgroundColor": "transparent",
            "overflow": "hidden",
            "position": "relative",
          }
        }
      >
        <Image
          onLoad={[Function]}
          replaceTheme={[Function]}
          source={Object {}}
          style={
            Object {
              "bottom": 0,
              "height": 100,
              "left": 0,
              "opacity": 0.4,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 100,
            }
          }
          testID="RNE__Image"
          theme={
            Object {
              "colors": Object {
                "background": "#ffffff",
                "black": "#242424",
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "grey": "rgba(0, 0, 0, 0.54)",
                    "primary": "#2196f3",
                    "searchBg": "#dcdce1",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "default": Object {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "web": Object {
                    "error": "#ff190c",
                    "grey": "#393e42",
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#03dac4",
                "success": "#52c41a",
                "warning": "#faad14",
                "white": "#ffffff",
              },
              "mode": "light",
              "spacing": Object {
                "lg": 12,
                "md": 8,
                "sm": 4,
                "xl": 24,
                "xs": 2,
              },
            }
          }
          transitionDuration={360}
          updateTheme={[Function]}
        />
        <View
          accessibilityElementsHidden={true}
          collapsable={false}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#bdbdbd",
                "height": 100,
                "justifyContent": "center",
                "opacity": 0.4,
                "width": 100,
              }
            }
            testID="RNE__Image__placeholder"
          />
        </View>
        <View
          style={
            Object {
              "height": 100,
              "opacity": 0.4,
              "width": 100,
            }
          }
          testID="RNE__Image__children__container"
        />
      </View>
    </View>
    <Text
      accessibilityRole="text"
      style={
        Object {
          "color": "#242424",
          "marginLeft": "auto",
          "marginRight": "auto",
          "marginTop": 32,
        }
      }
    >
      Add new tasks via clicking the plus sign (+)
    </Text>
  </View>
</RCTSafeAreaView>
`;
